common/azure_ps_context_credential.go

```diff
@@ -0,0 +1,186 @@
+//
+// [Azure Developer CLI]: https://learn.microsoft.com/azure/developer/azure-developer-cli/overview
+type PowershellContextCredential struct {
+	mu   *sync.Mutex
+	opts PowershellContextCredentialOptions
+}
+
+func NewPowershellContextCredential(options *PowershellContextCredentialOptions) (*PowershellContextCredential, error) {
+	cp := PowershellContextCredentialOptions{}
+	if options != nil {
+		cp = *options
+	}
+	if cp.TenantID != "" && !validTenantID(cp.TenantID) {
+		return nil, errors.New("invalid tenant id")
+	}
+	if cp.tokenProvider == nil {
+		cp.tokenProvider = defaultAzdTokenProvider
+	}
+	return &PowershellContextCredential{mu: &sync.Mutex{}, opts: cp}, nil
+}
+
+// GetToken requests a token from the Azure Developer CLI. This credential doesn't cache tokens, so every call invokes azd.
+// This method is called automatically by Azure SDK clients.
+func (c *PowershellContextCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	at := azcore.AccessToken{}
+	if len(opts.Scopes) != 1 {
+		return at, errors.New(credNamePSContext + ": GetToken() exactly one scope")
+	}
+
+	tenant, err := resolveTenant(c.opts.TenantID, opts.TenantID, credNamePSContext, nil)
+	if err != nil {
+		return at, err
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	b, err := c.opts.tokenProvider(ctx, opts.Scopes[0], tenant)
+	if err == nil {
+		at, err = c.createAccessToken(b)
+	}
+	if err != nil {
+		return at, err
+	}
+	//msg := fmt.Sprintf("%s.GetToken() acquired a token for scope %q", credNamePSContext, strings.Join(opts.Scopes, ", "))
+	return at, nil
+}
+
+// We ignore resource because PS does not support all Resources. Disk scope is not supported
+// and we are here only with Storage scope
+var defaultAzdTokenProvider PSTokenProvider = func(ctx context.Context, _ string, tenantID string) ([]byte, error) {
+	// set a default timeout for this authentication iff the application hasn't done so already
+	var cancel context.CancelFunc
+	if _, hasDeadline := ctx.Deadline(); !hasDeadline {
+		ctx, cancel = context.WithTimeout(ctx, 10 * time.Minute)
+		defer cancel()
+	}
+
+	const StorageResourceName = "Storage"
+	commandLine := "Get-AzAccessToken -ResourceTypeName " + StorageResourceName + " | ConvertTo-Json"
+	if tenantID != "" {
+		commandLine += " -TenantId " + tenantID
+	}
+	cliCmd := exec.CommandContext(ctx, "powershell", commandLine)
+	cliCmd.Env = os.Environ()
+	var stderr bytes.Buffer
+	cliCmd.Stderr = &stderr
+
+	output, err := cliCmd.Output()
+	if err != nil {
+		msg := stderr.String()
+		var exErr *exec.ExitError
+		if errors.As(err, &exErr) && exErr.ExitCode() == 127 || strings.HasPrefix(msg, "Not found") {
+			msg = "Powershell path not found"
+		}
+		if msg == "" {
+			msg = err.Error()
+		}
+		return nil, errors.New(credNamePSContext + msg)
+	}
+
+	return output, nil
+}
+
+func (c *PowershellContextCredential) createAccessToken(tk []byte) (azcore.AccessToken, error) {
```

Comment:
can we add UT for some of these methods?

Reply:
Did you add tests for some of these methods?

Reply:
The test covers entire workflow.

Reply:
KMN. Thanks for getting this. I had not checked in the file. I've checked in now, need to get some environment vars correct in CI.

Reply:
great! thanks for double checking that. I thought I was going crazy for a minute